File: sdat2img.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type block_command struct {
	command string
	blocks  []block_range
}

type block_range struct {
	a int
	b int
}

const BLOCK_SIZE = 4096

func print_help() {
	fmt.Println("./sdat2img <system.transfer.list> <system.new.dat> [system.img]")
}

func xmax(r []block_range) int {
	n := 0
	for _, cur := range r {
		if cur.a > n || cur.b > n {
			if cur.a > cur.b {
				n = cur.a
			} else {
				n = cur.b
			}
		}
	}
	return n * BLOCK_SIZE
}

func rangeset(s string) []block_range {
	var r []block_range
	src_set := strings.Split(s, ",")
	var num_set []int

	for _, c := range src_set {
		n, _ := strconv.Atoi(c)
		num_set = append(num_set, n)
	}

	if len(num_set) != num_set[0]+1 {
		fmt.Fprintln(os.Stderr, "Error on parsing following data to rangeset:\n", s)
		os.Exit(1)
	}

	for i := 1; i < len(num_set); i += 2 {
		r = append(r, block_range{num_set[i], num_set[i+1]})
	}

	return r
}

func parse_transfer_list_file(flist io.ReadCloser) (int, int, []block_command) {
	var version, new_blocks int
	var commands []block_command

	isdigit := func(s string) bool {
		_, err := strconv.Atoi(s)
		return err == nil
	}

	scanner := bufio.NewScanner(flist)

	scanner.Scan()
	version, _ = strconv.Atoi(scanner.Text())
	scanner.Scan()
	new_blocks, _ = strconv.Atoi(scanner.Text())

	if version >= 2 {
		scanner.Scan()
		scanner.Scan()
	}

	var line string
	for scanner.Scan() {
		line = scanner.Text()

		s := strings.Split(line, " ")
		cmd := s[0]

		if cmd == "erase" || cmd == "new" || cmd == "zero" {
			commands = append(commands, block_command{cmd, rangeset(s[1])})
		} else {
			if !isdigit(cmd[0:1]) {
				fmt.Fprintln(os.Stderr, "Command ", cmd, " is not valid.")
				flist.Close()
				os.Exit(1)
			}
		}

	}

	return version, new_blocks, commands
}

func main() {

	var transfer_list_file, new_dat_file, output_file string
	var err error

	if len(os.Args) == 4 {
		transfer_list_file = os.Args[1]
		new_dat_file = os.Args[2]
		output_file = os.Args[3]
	} else {
		print_help()
		os.Exit(1)
	}

	flist, err := os.OpenFile(transfer_list_file, os.O_RDONLY, 0644)
	if err != nil {
		fmt.Println("Error: ", err.Error())
		return
	}
	defer flist.Close()
	finfo, err := flist.Stat()
	if finfo.IsDir() {
		fmt.Println("Error: transfer.list is not a file but dir.")
		return
	}

	var version int
	var commands []block_command

	version, _, commands = parse_transfer_list_file(flist)

	if version == 1 {
		fmt.Println("Android Lollipop 5.0 detected!")
	} else if version == 2 {
		fmt.Println("Android Lollipop 5.1 detected!")
	} else if version == 3 {
		fmt.Println("Android Marshmallow 6.x detected!")
	} else if version == 4 {
		fmt.Println("Android Nougat 7.x / Oreo 8.x detected!")
	} else {
		fmt.Println("Unknown Android version!")
	}

	fdat, err := os.OpenFile(new_dat_file, os.O_RDONLY, 0644)
	if err != nil {
		fmt.Println("Error: ", err.Error())
		return
	}
	defer fdat.Close()
	finfo, err = fdat.Stat()
	if finfo.IsDir() {
		fmt.Println("Error: new.dat is not a file but dir.")
		return
	}

	fout, err := os.Create(output_file)
	if err != nil {
		fmt.Println("Error: ", err.Error())
		return
	}
	defer fout.Close()

	var all_block_sets []block_range
	for _, cur := range commands {
		all_block_sets = append(all_block_sets, cur.blocks...)
	}

	max_file_size := xmax(all_block_sets)
	var begin, end, block_count int
	buf := make([]byte, 4096)

	//fdat_reader := bufio.NewReader(fdat)
	//fout_writer := bufio.NewWriter(fout)

	for _, command := range commands {
		if command.command == "new" {
			for _, block := range command.blocks {
				begin = block.a
				end = block.b
				block_count = end - begin

				fmt.Println("Copying ", block_count, " blocks into position ", begin, "...")

				fout.Seek(int64(begin*BLOCK_SIZE), io.SeekStart)

				for block_count > 0 {
					fdat.Read(buf)
					fout.Write(buf)
					block_count--
				}
			}
		} else {
			fmt.Println("Skipping command ", command.command, "...")
		}
	}

	pos, _ := fout.Seek(0, io.SeekCurrent)
	if pos < int64(max_file_size) {
		fout.Truncate(int64(max_file_size))
	}

	abs, _ := filepath.Abs(fout.Name())
	fmt.Println("Done! Output image: ", abs)
}

File: sdat2img.cpp
// LICENISE         : GPLv3
// Author           : affggh
// Original script  : https://github.com/xpirt/sdat2img

#define VERSION 1.2
#define BLOCK_SIZE 4096

// 32bit file max only 2Gb
// so we use 64 bits offset
#ifndef _FILE_OFFSET_BITS
#define _FILE_OFFSET_BITS 64
#endif

#include <iostream>
#include <vector>
#include <sstream>
#include <fstream>
#include <cstring>

#include <sys/stat.h>
#include <sys/types.h>
#include <sys/fcntl.h>

using namespace std;

struct block_range
{
    streamoff a;
    streamoff b;
};

struct block_command
{
    string              command;
    vector<block_range> blocks;
};

streamoff max(vector<block_range> &n)
{
    streamoff max_num = 0;
    for (const auto &i : n) {
       if (i.a > max_num || i.b > max_num) {
            max_num = (i.a > i.b) ? i.a : i.b;
       }
    }
    return max_num * BLOCK_SIZE;
}

class SDAT2IMG 
{
private:
    ifstream flist, fdat;
    ofstream fout;

public:
    errno_t etype;

    SDAT2IMG(const char *transfer_list_file, const char *new_dat_file, const char *output_file) 
    {
        if (!(access(transfer_list_file, F_OK) == 0)){
            cerr << "Error: file " << transfer_list_file << "does not exist!" << endl;
            etype = EIO;
            return;
        }

        if (!(access(new_dat_file, F_OK) == 0)) {
            cerr << "Error: file " << new_dat_file << "does not exist!" << endl;
            etype = EIO;
            return;
        }

        struct stat st;
        stat(transfer_list_file, &st);
        if (S_ISDIR(st.st_mode)) {
            cerr << "Error: " << transfer_list_file << "is a dir!" << endl;
            etype = EIO;
            return;
        }

        stat(new_dat_file, &st);
        if (S_ISDIR(st.st_mode)) {
            cerr << "Error: " << new_dat_file << "is a dir!" << endl;
            etype = EIO;
            return;
        }

        flist.open(transfer_list_file);
        if (!flist) {
            cerr << "Error: " << "could not open transfer list file." << endl;
            etype = EIO;
            return;
        }

        fdat.open(new_dat_file, ios::binary);
        if (!fdat) {
            cerr << "Error: " << "could not open new.dat file." << endl;
            etype = EIO;
            return;
        }

        fout.open(output_file, ios::binary | ios::trunc);
        if (!fout) {
            cerr << "Error: " << "could not open output file." << endl;
            etype = EIO;
            return;
        }
    }

    vector<block_range> split_blocks(string blocks) {
        string line;
        istringstream iss(blocks);
        auto num_len = 0ull;
        vector<streamoff> b;
        vector<block_range> r;

        while(getline(iss, line, ',')) {
            if (num_len == 0) {
                num_len = atoi(line.c_str());
                continue;
            }
            b.push_back(atoi(line.c_str()));
        }

        if (num_len != b.size()) {
            cerr << "Error on parsing following data to rangeset:" << endl << blocks << endl;
            etype = EIO;
            return r;
        }

        for (auto i=0ull; i<b.size(); i+=2) {
            r.push_back(block_range{b[i], b[i+1]});
        }

        return r;
    }

    int parse_transfer_list_file(int &version, int &new_blocks, vector<block_command> &c) {
        int ret = 0;

        istringstream iss;
        string line, cmd, b;

        getline(flist, line);
        version = atoi(line.c_str());

        getline(flist, line);
        new_blocks = atoi(line.c_str());

        if (version >= 2) {
            getline(flist, line);
            getline(flist, line);
        }

        while(getline(flist, line)) {
            iss.clear();
            iss.str(line);
            iss >> cmd;
            if (cmd == "erase" || cmd == "new" || cmd == "zero") {
                iss >> b;
                c.push_back(block_command{cmd, split_blocks(b)});
                if (etype) {
                    ret = etype;
                    break;
                }
            } else {
                cerr << "Command " << cmd << " is not valid." << endl;
                ret = 1;
                break;
            }
        }

        return ret;
    }

    ~SDAT2IMG()
    {
        if (flist)
            flist.close();
        if (fdat)
            fdat.close();
        if (fout)
            fout.close();
    }

    int run(void)
    {
        int                     version, new_blocks;
        vector<block_command>   commands;
        vector<block_range>     all_block_sets;
        streamoff                 max_file_size = 0;
        char                    buf[BLOCK_SIZE];
        streamoff               begin, end, block_count;

        parse_transfer_list_file(version, new_blocks, commands);

        switch (version) {
            case 1:
                cout << "Android Lollipop 5.0 detected!" << endl;
                break;
            case 2:
                cout << "Android Lollipop 5.1 detected!" << endl;
                break;
            case 3:
                cout << "Android Marshmallow 6.x detected!" << endl;
                break;
            case 4:
                cout << "Android Nougat 7.x / Oreo 8.x detected!" << endl;
                break;
            default:
                cout << "Unknown Android version!" << endl;
                break;
        }

        for (const auto &i : commands) {
            for (const auto &b : i.blocks) {
                all_block_sets.push_back(b);
            }
        }

        max_file_size = max(all_block_sets);

        for (const auto &c : commands) {
            if (c.command == "new") {
                for (const auto &block : c.blocks) {
                    begin = block.a;
                    end = block.b;
                    block_count = end - begin;

                    cout << "Copying " << block_count <<" blocks into position " << begin << "..." << fout.tellp() << endl;;

                    fout.seekp(begin*BLOCK_SIZE, ios::beg);

                    while (block_count > 0) {
                        fdat.read(buf, sizeof(buf));
                        fout.write(buf, sizeof(buf));
                        block_count--;
                    }
                }
            } else {
                cout << "Skipping command " << c.command << "..." << endl;
            }
        }

        if (fout.tellp() < max_file_size) {
            fout.seekp(max_file_size, ios::beg);
            fout.flush();
        }
        return 0;
    }
};

void help(void) 
{
    cout << "Usage: ./sdat2img [system.transfer.list] [system.new.dat] <system.img>" << endl;
}

int main(int argc, char** argv) 
{
    int ret = 0;
    char *transfer_list_file, *new_dat_file, *output_file = nullptr;

    cout << "sdat2img cpp version: " << VERSION << endl;
    if (argc < 3 || argc > 4) {
        help();
        return 1;
    }

    if (argc == 3) {
        output_file = strdup("system.img");
    } else if (argc == 4) {
        output_file = strdup(argv[3]);
    }
    transfer_list_file = strdup(argv[1]);
    new_dat_file = strdup(argv[2]);

    SDAT2IMG sdat2img(transfer_list_file, new_dat_file, output_file);
    if (sdat2img.etype) {
        goto error;
    }

    ret = sdat2img.run();
    if (!ret)
        goto success;

error:
    cerr << "Error: something went worng!" << endl;
    ret = sdat2img.etype;

success:
    if (output_file)
        free(output_file);
    if(transfer_list_file)
        free(transfer_list_file);
    if (new_dat_file)
        free(new_dat_file);

    return ret;
}
